src/util/hash-list.h

```diff
@@ -57,34 +57,82 @@ template<class I, class T> class HashList {
 
   /// Constructor takes no arguments.
   /// Call SetSize to inform it of the likely size.
-  HashList();
+  HashList() : list_head_(nullptr), bucket_list_tail_(boundary_tag_),
+               hash_size_(0), freed_head_(nullptr) {}
 
   /// Clears the hash and gives the head of the current list to the user;
   /// ownership is transferred to the user (the user must call Delete()
   /// for each element in the list, at his/her leisure).
-  Elem *Clear();
+  Elem *Clear() noexcept {
+    // Clears the hashtable and gives ownership of the currently contained list
+    // to the user.
+    for (size_t cur_bucket = bucket_list_tail_; cur_bucket != boundary_tag_;
+         cur_bucket = buckets_[cur_bucket].prev_bucket) {
+      buckets_[cur_bucket].last_elem = nullptr;  // this is how we indicate "empty".
+    }
+
+    bucket_list_tail_ = boundary_tag_;
+    Elem *ans = list_head_;
+    list_head_ = nullptr;
+    return ans;
+  }
 
   /// Gives the head of the current list to the user.  Ownership retained in the
   /// class.  Caution: in December 2013 the return type was changed to const
   /// Elem* and this function was made const.  You may need to change some types
   /// of local Elem* variables to const if this produces compilation errors.
-  const Elem *GetList() const;
+  const Elem *GetList() const noexcept { return list_head_; }
 
   /// Think of this like delete().  It is to be called for each Elem in turn
   /// after you "obtained ownership" by doing Clear().  This is not the opposite
   /// of. Insert, it is the opposite of New.  It's really a memory operation.
-  inline void Delete(Elem *e);
+  inline void Delete(Elem *e) noexcept {
+    e->tail = freed_head_;
+    freed_head_ = e;
+  }
 
   /// This should probably not be needed to be called directly by the user.
-  /// Think of it as opposite
-  /// to Delete();
-  inline Elem *New();
+  /// Think of it as opposite to Delete();
+  inline Elem *New() noexcept {
+    if (!freed_head_) {
+      Elem *tmp = new Elem[allocate_block_size_];
```

Comment:
This cannot be marked noexcept because it is doing a memory allocation. It can throw std::bad_alloc.

Reply:
You are right.